cmd/tidb-dashboard/main.go

```diff
@@ -155,65 +151,49 @@ func getContext() context.Context {
 }
 
 func main() {
-	_ = godotenv.Load()
-
 	// Flushing any buffered log entries
 	defer log.Sync() //nolint:errcheck
 
 	cliConfig := NewCLIConfig()
 	ctx := getContext()
 
-	store := dbstore.MustOpenDBStore(cliConfig.CoreConfig)
-	defer store.Close() //nolint:errcheck
+	if cliConfig.EnableDebugLog {
+		log.SetLevel(zapcore.DebugLevel)
+	}
 
-	etcdProvider, err := pd.NewLocalEtcdClientProvider(cliConfig.CoreConfig)
+	listenAddr := fmt.Sprintf("%s:%d", cliConfig.ListenHost, cliConfig.ListenPort)
+	listener, err := net.Listen("tcp", listenAddr)
 	if err != nil {
-		_ = store.Close()
-		log.Fatal("Cannot create etcd client", zap.Error(err))
+		log.Fatal("Dashboard server listen failed", zap.String("addr", listenAddr), zap.Error(err))
 	}
 
-	tidbForwarder := tidb.NewForwarder(tidb.NewForwarderConfig(cliConfig.CoreConfig.TLSConfig), etcdProvider)
-	// FIXME: Handle open error
-	tidbForwarder.Open()        //nolint:errcheck
-	defer tidbForwarder.Close() //nolint:errcheck
+	s := apiserver.NewService(
+		cliConfig.CoreConfig,
+		apiserver.StoppedHandler,
+		func(cfg *config.Config, httpClient *http.Client, etcdClient *clientv3.Client) *keyvisualregion.PDDataProvider {
+			return &keyvisualregion.PDDataProvider{
+				FileStartTime:  cliConfig.KVFileStartTime,
+				FileEndTime:    cliConfig.KVFileEndTime,
+				PeriodicGetter: keyvisualinput.NewAPIPeriodicGetter(cliConfig.CoreConfig.PDEndPoint, httpClient),
+				EtcdClient:     etcdClient,
+			}
+		},
+	)
+	if err := s.Start(ctx); err != nil {
+		log.Fatal("Can not start server", zap.Error(err))
+	}
+	defer s.Stop(context.Background()) //nolint:errcheck
 
-	httpClient := http2.NewHTTPClientWithConf(cliConfig.CoreConfig)
+	r := gin.New()
+	r.Use(gin.Recovery())
+	apiserver.Register(&r.RouterGroup, s)
 
-	// key visual
-	remoteDataProvider := &keyvisualregion.PDDataProvider{
-		FileStartTime:  cliConfig.KVFileStartTime,
-		FileEndTime:    cliConfig.KVFileEndTime,
-		PeriodicGetter: keyvisualinput.NewAPIPeriodicGetter(cliConfig.CoreConfig.PDEndPoint, httpClient),
-		EtcdProvider:   etcdProvider,
-		Store:          store,
-	}
-	keyvisualService := keyvisual.NewService(ctx, cliConfig.CoreConfig, remoteDataProvider, httpClient)
-	keyvisualService.Start()
-	defer keyvisualService.Close()
-
-	services := &apiserver.Services{
-		Store:         store,
-		KeyVisual:     keyvisualService,
-		TiDBForwarder: tidbForwarder,
-		EtcdProvider:  etcdProvider,
-		HTTPClient:    httpClient,
-	}
 	mux := http.DefaultServeMux
-	mux.Handle("/dashboard/", http.StripPrefix("/dashboard", uiserver.Handler()))
-	mux.Handle("/dashboard/api/", apiserver.Handler("/dashboard/api", cliConfig.CoreConfig, services))
+	mux.Handle("/dashboard/", s.NewStatefulHandler(http.StripPrefix("/dashboard", uiserver.Handler())))
```

Comment:
When this is running standlone, maybe it is not very necessary to reject the UI since it will not be stopped.. Then `s.NewStatefulHandler` can be removed.